src/workerd/io/io-context.h

```diff
@@ -983,15 +983,22 @@ jsg::PromiseForResult<Func, T, true> IoContext::awaitIoImpl(
             // before the inner promise.
             func = kj::fwd<Func>(func)]
             (ExceptionOr<T>&& exceptionOrT) mutable {
+    struct FuncResultPair {
+      // It's important that `exceptionOrT` is destroyed before `Func`. Lambda captures are
+      // destroyed in unspecified order, so we wrap them in a struct to make it explicit.
+      Func func;
+      ExceptionOr<T> exceptionOrT;
+    };
+
     return run([resolver = kj::mv(resolver),
-                exceptionOrT = kj::mv(exceptionOrT),
-                maybeAsyncContext = kj::mv(maybeAsyncContext),
-                func = kj::fwd<Func>(func)](Worker::Lock& lock) mutable {
+                funcResultPair = FuncResultPair { kj::fwd<Func>(func), kj::mv(exceptionOrT) },
```

Comment:
I suppose this could also be done by explicitly moving the func and the value into the body of the function, assuming the function doesn't run twice.

Reply:
Oh but wait if the function is never called the func and value *won't* be moved into the body and so they could still fail to destruct properly.

Reply:
Right, it's the never-called case that's an issue here.